File: starfightersquadrons/src/main/java/org/vizzini/starfightersquadrons/ShipDestroyedAction.java
package org.vizzini.starfightersquadrons;

import java.util.ArrayList;
import java.util.List;

import org.vizzini.core.InputValidator;
import org.vizzini.core.game.Action;

/**
 * Provides a ship destroyed action for Starfighter Squadrons.
 */
public final class ShipDestroyedAction implements Action
{
    /** Environment. */
    private final SSEnvironment environment;

    /** From position. */
    private final SSPosition fromPosition;

    /** Token. */
    private final SSToken token;

    /**
     * Construct this object.
     *
     * @param environment Environment.
     * @param token Token.
     * @param fromPosition From position.
     */
    @SuppressWarnings("hiding")
    public ShipDestroyedAction(final SSEnvironment environment, final SSToken token,
            final SSPosition fromPosition)
    {
        InputValidator.validateNotNull("environment", environment);
        InputValidator.validateNotNull("token", token);
        InputValidator.validateNotNull("fromPosition", fromPosition);

        this.environment = environment;
        this.token = token;
        this.fromPosition = fromPosition;
    }

    @Override
    public boolean doIt()
    {
        final TargetLock attackerTargetLock = token.getAttackerTargetLock();

        if (attackerTargetLock != null)
        {
            TargetLock.freeInstance(attackerTargetLock);
        }

        final List<TargetLock> defenderTargetLocks = new ArrayList<TargetLock>(token.getDefenderTargetLocks());

        for (final TargetLock defenderTargetLock : defenderTargetLocks)
        {
            TargetLock.freeInstance(defenderTargetLock);
        }

        // Return the damage cards.
        environment.discardAllDamage(token.getDamages());
        environment.discardAllDamage(token.getCriticalDamages());

        environment.removeToken(fromPosition);
        environment.fireShipDestroyedActionPropertyChange(null, this);

        return true;
    }

    @Override
    public SSAgent getAgent()
    {
        return token.getAgent();
    }

    @Override
    public SSEnvironment getEnvironment()
    {
        return environment;
    }

    /**
     * @return the fromPosition
     */
    public SSPosition getFromPosition()
    {
        return fromPosition;
    }

    /**
     * @return the token
     */
    public SSToken getToken()
    {
        return token;
    }

    @Override
    public boolean undoIt()
    {
        environment.placeToken(fromPosition, token);
        environment.fireShipDestroyedActionPropertyChange(null, this);

        return true;
    }
}

File: game/starfightersquadrons/src/main/js/process/ShipDestroyedAction.js
define(["process/Action", "process/TargetLock"],
   function(Action, TargetLock)
   {
      "use strict";

      function ShipDestroyedAction(environment, token, fromPosition)
      {
         InputValidator.validateNotNull("environment", environment);
         InputValidator.validateNotNull("token", token);

         this.environment = function()
         {
            return environment;
         };

         this.token = function()
         {
            return token;
         };

         this.fromPosition = function()
         {
            return fromPosition;
         };

         this.doIt = function()
         {
            LOGGER.trace("ShipDestroyedAction.doIt() start");

            var store = environment.store();
            TargetLock.removeAllTargetLocks(store, token);

            // Return the damage cards.
            if (token.tokenFore && token.tokenAft)
            {
               environment.discardAllDamage(token.tokenFore().damageKeys());
               environment.discardAllDamage(token.tokenFore().criticalDamageKeys());
               environment.discardAllDamage(token.tokenAft().damageKeys());
               environment.discardAllDamage(token.tokenAft().criticalDamageKeys());
            }
            else
            {
               environment.discardAllDamage(token.damageKeys());
               environment.discardAllDamage(token.criticalDamageKeys());
            }

            if (fromPosition)
            {
               environment.removeToken(fromPosition);
            }
            store.dispatch(Action.setUserMessage("Ship destroyed: " + token));

            LOGGER.trace("ShipDestroyedAction.doIt() end");
         };
      }

      return ShipDestroyedAction;
   });
